src/clj/puppetlabs/http/client/async.clj

```diff
@@ -117,11 +124,68 @@
       (.setAs (clojure-response-body-type->java opts))
       (.setBody (:body opts))
       (.setDecompressBody (clojure.core/get opts :decompress-body true))
-      (.setHeaders (:headers opts))))
+      (.setHeaders (:headers opts))
+      (.setMetricId (parse-metric-id opts))))
+
+(schema/defn get-mean :- schema/Num
+  [timer :- Timer]
+  (->> timer
+       .getSnapshot
+       .getMean
+       (.toMillis TimeUnit/NANOSECONDS)))
+
+(defn get-metric-data
+  [timer metric-id]
+  (let [count (.getCount timer)
+        mean (get-mean timer)
+        aggregate (* count mean)]
+    {:count count
+     :mean mean
+     :aggregate aggregate
+     :metric-id metric-id}))
+
+(defn get-metrics-data
+  [timers]
+  (reduce (fn [acc [metric-id timer]]
+            (assoc acc metric-id (get-metric-data timer metric-id)))
+          {} timers))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Public
 
+(schema/defn ^:always-validate get-client-metrics :- (schema/maybe common/Metrics)
+  "Returns the http client-specific metrics from the metric registry."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (when metric-registry
+     (into {} (Metrics/getClientMetrics metric-registry))))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
+   (when metric-registry
+     (cond
+       (:verb metric-filter) (into {} (Metrics/getClientMetricsWithUrlAndVerb
+                                       metric-registry
+                                       (:url metric-filter)
+                                       (:verb metric-filter)
+                                       Metrics$MetricType/BYTES_READ))
+       (:url metric-filter) (into {} (Metrics/getClientMetricsWithUrl
+                                      metric-registry
+                                      (:url metric-filter)
+                                      Metrics$MetricType/BYTES_READ))
+       (:metric-id metric-filter) (into {} (Metrics/getClientMetricsWithMetricId
+                                            metric-registry
+                                            (into-array (:metric-id metric-filter)) Metrics$MetricType/BYTES_READ))
+       :else (throw (IllegalArgumentException. "Not an allowed metric filter."))))))
+
+(schema/defn ^:always-validate get-client-metrics-data :- (schema/maybe common/MetricsData)
+  "Returns a map of metric-id to metric data summary."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (let [timers (get-client-metrics metric-registry)]
+     (get-metrics-data timers)))
```

Comment:
I notice that these fns return an empty map in the case where the client was created without a MetricsRegistry.

It might be worth considering whether they should instead return `nil`, or even throw an exception.  Not 100% sure what I think is the best out of those three options, just food for thought.